Add Collision Monitor documentaiton pages (#333)
ros-planning/navigation.ros.org: configuration/packages/configuring-collision-monitor.rst

.. _configuring_collision_monitor:

Collision Monitor
#################

> Collision Monitor - is an independent layer in Nav2 providing an additional level of robot safety assurance.

Reviewer: Update with README's requested introduction changes

Reviewer: In general below, apply changes I made to the readme here

It allows performing robot collision avoidance by monitoring the obstacles in surrounding environments laying in collision proximity to the robot.

See the package's ``README`` for more complete information.

Features
********

The following models of safety behaviors are employed by Collision Monitor:

- **Stop model**: Define a safety area surrounding the robot and a point threshold. If more that ``max_points`` appear inside this area, stop the robot until the obstacles will disappear.

> - **Slowdown model**: Define a safety area around the robot and slow the maximum speed for a ``slowdown_ratio``, if more than ``max_points`` points will appear inside the area.

Reviewer: ```suggestion
- **Slowdown model**: Define a safety area around the robot and slow the maximum speed by a ``slowdown_ratio``, if more than ``max_points`` points will appear inside the area.
```

  double         0.1
  ============== =============================

  Description
    Time with which to post-date the transform that is published, to indicate that this transform is valid into the future.

:source_timeout:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  double         2.0
  ============== =============================

  Description:
    Maximum time interval in which source data is considered as valid.

:stop_pub_timeout:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  double         1.0
  ============== =============================

  Description:
    Timeout, after which zero-velocity ceases to be published. It could be used for other overrode systems outside Nav2 are trying to bring the robot out of a state close to a collision, or to allow a standing robot to go into sleep mode.

:polygons:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  vector<string> N/A
  ============== =============================

  Description:

>     List of safety area shapes (stop/slowdown bounding boxes, footprint, approach circle, etc...). Causes an error, if not specialized.

Reviewer: Zones, remove the phrase "safety area" from all docs, we should avoid the word "safety"

Author: Remove all phrases in docs and README (changed to "zones" or similar)

:observation_sources:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  vector<string> N/A
  ============== =============================

  Description:

>     List of data sources (llaser scanners, pointclouds, etc...). Causes an error, if not specialized.

Reviewer: ```suggestion
    List of data sources (laser scanners, pointclouds, etc...). Causes an error, if not specialized.
```

Polygons parameters
===================

``<polygon name>`` is the corresponding polygon name ID selected for this type.

:``<polygon_name>``.type:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  string         N/A
  ============== =============================

  Description:
    Type of polygon shape. Available values are ``polygon``, ``circle``. Causes an error, if not specialized.

:``<polygon_name>``.points:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  vector<double> N/A
  ============== =============================

  Description:

>     Polygon vertexes, listed in ``{p1.x, p1.y, p2.x, p2.y, p3.x, p3.y, ...}`` format. Used for ``polygon`` type. Minimum 3 points for a triangle polygon. Causes an error, if not specialized.

Reviewer: Give an example box

Author: [Done](https://github.com/ros-planning/navigation.ros.org/pull/333/files#diff-60e7d74687df0b5a07a3d91dc411c686e266df32edbdb653cf6e2ffc2674bc7cR175)

:``<polygon_name>``.radius:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  double         N/A
  ============== =============================

  Description:
    Circle radius. Used for ``circle`` type. Causes an error, if not specialized.

:``<polygon_name>``.action_type:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  string         N/A
  ============== =============================

  Description:

>     Safety behavior model. Available values are ``stop``, ``slowdown``, ``approach``. Causes an error, if not specialized.

Reviewer: ```suggestion
    Zone model. Available values are ``stop``, ``slowdown``, ``approach``. Causes an error, if not specialized.
```

Author: [Changed](https://github.com/ros-planning/navigation.ros.org/pull/333/files#diff-60e7d74687df0b5a07a3d91dc411c686e266df32edbdb653cf6e2ffc2674bc7cR197) to "Zone behavior model"

:``<polygon_name>``.max_points:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  int            3
  ============== =============================

  Description:

>     Maximum number of points to enter inside polygon to be ignored (w/o causing an action).

Reviewer: ```suggestion
    Number of data readings within a zone to trigger the action
```

Author: The question about the change: should this parameter to be a number of a points, *starting from* an action will be triggered, or maximum number of points *to be ignored before* an action? According to the initial task description from the ticket 1899:

> Define a bounding box and a point threshold and if > N points appear in this box in front of the robot, stop it. Usually with a larger bounding box in front of the stop box to warn of a collision and slow the maximum speeds.

It is second definition. Means, that if `> max_points` will enter inside polygon, the action will be triggered.

If read the proposed definition,
> Number of data readings within a zone to trigger the action

it should be `>= max_points` to trigger an action.

To fit initial conditions, making the following change in docs instead:
```
- Maximum number of points to enter inside polygon to be ignored (w/o causing an action).
+ Maximum number of data readings within a zone to not trigger the action
```
Please let me know, if this number should have another meaning.

  ============== =============================
  Type           Default
  -------------- -----------------------------
  double         0.02
  ============== =============================

  Description:
    Time iteration step for robot movement simulation during collision prediction. Lower values mean lower prediction accuracy but better performance. Applicable for ``approach`` action type.

:``<polygon_name>``.visualize:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  bool           False
  ============== =============================

  Description:
    Whether to publish the polygon in a separate topic.

:``<polygon_name>``.polygon_pub_topic:

  ============== =============================
  Type           Default
  -------------- -----------------------------
  string         <polygon_name>
  ============== =============================

  Description:
    Topic name to publish a polygon to. Used only if ``visualize`` is true.

:``<polygon_name>``.footprint_topic:

  ============== =============================
  Type           Default
  -------------- -----------------------------

>   string         "footprint"

Reviewer: I don't think that's a sensible default, use the topic the local costmap publishes with the footprint

Author: [Changed](https://github.com/ros-planning/navigation.ros.org/pull/333/files#diff-60e7d74687df0b5a07a3d91dc411c686e266df32edbdb653cf6e2ffc2674bc7cR270) to "local_costmap/published_footprint"
